server/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java

```diff
@@ -232,7 +228,11 @@ public Failure(String index, String type, String id, Exception cause, RestStatus
          */
         public Failure(StreamInput in) throws IOException {
             index = in.readString();
-            type = in.readString();
+            if (in.getVersion().before(Version.V_8_0_0)) {
+                in.readString();
+                // can't make an assertion about type names here because too many tests still set their own
+                // types bypassing various checks
```

Comment:
How many tests do this? I'm wondering if we need to make a separate change to make all the tests use the typeless APIs and backport it to 7.x so we can have these checks?

Reply:
I'll re-add it in and change the relevant tests.

Reply:
OK, so the reason that we can't assert the type here is indeed for backwards compatibility; it's a failure message, and there are certain circumstances in which a user can supply information which results in a different type being returned here: specifically, if a user tries to index documents with different type names, then we return an error message saying you can only have a single type, and the `type` field in this failure message will be set to one of those user-supplied names.